### PR TITLE
Support optional chunked reads from S3

### LINKED ISSUE #725: [s3] Ranged read http header leaves TO bytes blank, causing S3 filesystems to read full file length
#### Problem description
`smart_open` is effectively crashing SeaweedFS/S3 and Ceph/S3 filesystems when doing many small reads over large files (ex: 32k read on a 4GB file).

On a SeaweedFS/S3 filesystem (also demonstrated on Ceph/S3), using the code shown in the reproduce section below, I am reading a small segment of data (32k in this example) from a large file (4GB in this example). This simple request causes SeaweedFS to move the full file for just a 32k range read. It appears that this is expected behavior based on a reading of the protocol specifications. Notably `boto3` does not trigger the same behavior.

When I run the code and look at the HTTP traffic being generated we see the following GET request:

```
GET /hengenlab/CAF77/Neural_Data/highpass_750/Headstages_256_Channels_int16_2021-02-02_14-28-24.bin HTTP/1.1
Host: seaweed-filer.seaweedfs:8333
Accept-Encoding: identity
Range: bytes=0-
User-Agent: Boto3/1.24.59 Python/3.8.2 Linux/5.4.0-125-generic Botocore/1.27.59
X-Amz-Date: 20220921T204638Z
X-Amz-Content-SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
Authorization: AWS4-HMAC-SHA256 Credential=KL2PPBIZ4OYKR420C28D/20220921/us-east-1/s3/aws4_request, SignedHeaders=host;range;x-amz-content-sha256;x-amz-date, Signature=8009c4fdc85311977066c6988047a72658579e02c02b544fa8d48d8d8b9e8d57
amz-sdk-invocation-id: eace9bc1-a1d1-4244-84ee-1caa164bc294
amz-sdk-request: attempt=1
```

Notably `Range: bytes=0-` is our culprit. My assumption is that `smart_open` intends to open the file for streaming and read data from the stream as dictated by calls to `f.read(...)`.

When performing a ranged read with just `boto3` code the header looks like this:

```
GET /hengenlab/CAF77/Neural_Data/highpass_750/Headstages_256_Channels_int16_2021-02-02_14-28-24.bin?partNumber=1 HTTP/1.1
Host: seaweed-filer.seaweedfs:8333
Accept-Encoding: identity
Range: bytes=1-32768
User-Agent: Boto3/1.24.59 Python/3.8.2 Linux/5.4.0-125-generic Botocore/1.27.59
X-Amz-Date: 20220921T205137Z
X-Amz-Content-SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
Authorization: AWS4-HMAC-SHA256 Credential=KL2PPBIZ4OYKR420C28D/20220921/us-east-1/s3/aws4_request, SignedHeaders=host;range;x-amz-content-sha256;x-amz-date, Signature=5d3398396c217f4a87284479bc6bc947344c256a30552fe98b6057167d7143fb
amz-sdk-invocation-id: 489f165b-cb4d-4ca0-bc49-cc1e70618518
amz-sdk-request: attempt=1
```

Using `boto3` does not cause any issue. With smart_open the SeaweedFS/S3 filesystem is interpreting the lack of a `to-bytes` value as the full file. It is then moving the full (4 GB) data file from a volume server to an S3 server, where just 32k are passed to the end user job. This has the effect of very quickly oversaturating the network. 

The protocol specifications seem to agree that the behavior by SeaweedFS/S3 is the correct way to interpret this Range header. E.g. how can the filesystem know that the user won't need to read the whole file given this header.

> If the last-byte-pos value is absent, or if the value is greater than or equal to the current length of the entity-body, last-byte-pos is taken to be equal to one less than the current length of the entity- body in bytes.

https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35

#### Steps/code to reproduce the problem

This code triggers the full (4 GB) file transfer (internally, not to the end user application) for a small 32k ranged read.
```
import import smart_open

while True:
    with smart_open.open('s3://bucket/path/to/4gb/file.bin', 'rb') as f:
        b = f.read(32768)
        print(f'Read {len(b)}')
```

This `boto3` version of the code does not trigger the same issue:

```
import boto3

while True:
    obj = boto3.resource('s3', endpoint_url='https://swfs-s3.endpoint').Object('bucket', 'path/to/4gb/file.bin')
    stream = obj.get(Range='bytes=1-32768')['Body']
    res = stream.read()
    print(f'Read: {len(res)} bytes')
```

I'd like to open a discussion to determine how to properly interpret the S3 protocol and figure out whether an issue like this should be in the domain of the filesystem(s) or should be changed in `smart_open`.

#### Versions

```
>>> print(platform.platform())
Linux-5.19.0-76051900-generic-x86_64-with-glibc2.17
>>> print("Python", sys.version)
Python 3.8.10 (default, Jun  4 2021, 15:09:15) 
[GCC 7.5.0]
>>> print("smart_open", smart_open.__version__)
smart_open 6.0.0
```

### PR DESCRIPTION
**Pull request by batterseapower**
This commit adds a new `range_chunk_size` parameter to the S3 reader that allows reading large files in smaller chunks. This is useful when you only need to read small portions of large S3 files, as it prevents S3-compatible storage systems from queueing up the entire file internally.

Key changes:
- Added range_chunk_size parameter to smart_open.s3.open() and related classes
- Modified _SeekableRawReader to support chunked reading with proper boundary handling

When range_chunk_size is None (default), behavior is unchanged - single request for the whole file to minimize per-request costs on S3.

> Please **pick a concise, informative and complete title** for your PR.
> 
> The title is important because it will appear in [our change log](https://github.com/RaRe-Technologies/smart_open/blob/master/CHANGELOG.md).

### Motivation

> Please explain the motivation behind this PR.
> 
> If you're fixing a bug, link to the issue using a [supported keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue) like "Fixes #{issue_number}".
> 
> If you're adding a new feature, then consider opening a ticket and discussing it with the maintainers before you actually do the hard work.

Fixes #725, uses an alternative approach to the stale PR #734

### Tests

> If you're fixing a bug, consider [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development):
> 
> 1. Create a unit test that demonstrates the bug. The test should **fail**.
> 2. Implement your bug fix.
> 3. The test you created should now **pass**.
> 
> If you're implementing a new feature, include unit tests for it.
> 
> Make sure all existing unit tests pass.
> You can run them locally using:
> 
>     pytest tests
> 
> If there are any failures, please fix them before creating the PR (or mark it as WIP, see below).

- Added comprehensive test coverage including adversarial testing for retry logic
- Improved error handling for edge cases (negative offsets, empty files, etc.)

### Work in progress

> If you're still working on your PR, mark the PR as [draft PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request).
> 
> We'll skip reviewing it for the time being.
> 
> Once it's ready, mark the PR as "ready for review", and ping one of the maintainers (e.g. mpenkov).

### Checklist

> Before you mark the PR as "ready for review", please make sure you have:

- [x] Picked a concise, informative and complete title
- [x] Clearly explained the motivation behind the PR
- [x] Linked to any existing issues that your PR will be solving
- [x] Included tests for any new functionality
- [x] Run `python update_helptext.py` in case there are API changes
- [x] Checked that all unit tests pass

### Workflow

> Please avoid rebasing and force-pushing to the branch of the PR once a review is in progress.
> 
> Rebasing can make your commits look a bit cleaner, but it also makes life more difficult from the reviewer, because they are no longer able to distinguish between code that has already been reviewed, and unreviewed code.

### CODE REVIEW

**Comment by ddelange** [comment]
Hi @batterseapower 👋 

Thanks for your efforts! 

Does this supercede https://github.com/piskvorky/smart_open/pull/734 and thus fix https://github.com/piskvorky/smart_open/issues/725? 

Please have a read, I wasn't a maintainer at the time but from this diff it looks like you're in pretty deep so I'd love to have a second opinion from you on the tradeoffs they're discussing.

**Comment by ddelange** [comment]
fwiw, I'm not opposed to releasing smart_open v8.0.0 with a behaviour change here if we can simplify + improve the code as opposed to adding complexity to the already complex s3 module. 

e.g. the retry logic has improved in boto3 over the years and i feel we could simplify the s3 module a lot in a single sweep here
```py
client = boto3.session.Session().client("s3", config=botocore.client.Config(
    max_pool_connections=64,  # thread-safe client ref https://github.com/boto/boto3/blob/1.38.41/docs/source/guide/clients.rst?plain=1#L111
    tcp_keepalive=True,  # re-use across GET requests without too much overhead
    retries={"max_attempts": 6, "mode": "adaptive"}  # exponential backoff built-in nowadays
))  
```

it would be great if the entire `fp.read()` does a single GET regardless of current position, but anything else, like iterating over lines, can be chunked by default imo. boto3 also has [`iter_lines`](https://github.com/boto/botocore/blob/1.40.31/botocore/response.py#L150-L163) which is chunked by default (but on a single HTTP request iirc)

a related/adjacent issue is our buffering issue: https://github.com/piskvorky/smart_open/issues/712 which also calls for simplification

**Comment by batterseapower** [comment]
> Does this supercede #734 and thus fix #725?

Yes, that's the idea!

I think #712 is out of scope for now - this would require some kind of more elaborate caching mechanism. I'm just trying to do the most minimal change that fixes #725.

> it would be great if the entire `fp.read()` does a single GET regardless of current position, but anything else, like iterating over lines, can be chunked by default imo. 

This wouldn't be hard to implement, but I'm wary about changing the default behaviour without a strong reason. It's hard to say whether switching to chunked by default for downstream users would break them in some way, given the diversity of use cases and s3-compatible APIs that could be in use.

File: smart_open/s3.py [hunk 1]
```diff
@@ -605,31 +681,68 @@ def read(self, size=-1):
...
+                try:
+                    if size == -1:
+                        binary = self._body.read()
+                    else:
+                        binary = self._body.read(size - bytes_read)
+                except (
+                    ConnectionResetError,
+                    botocore.exceptions.BotoCoreError,
+                    urllib3.exceptions.HTTPError,
+                ) as err:
```
[the remark this answers is not in the document]
Comment (batterseapower) [reply smart_open/s3.py:710]: Possibly, yes. The current approach is fairly readable though. We aren't doing anything very complicated with e.g. exponential backoff so no strong reason to migrate to tenacity IMHO (though I like tenacity and use it quite a bit)

**Comment by ddelange** [comment]
> Essentially analogous to multipart_upload=True which toggles between SinglepartWriter and MultipartWriter.

If these are going to be living next to each other, how about extending that analogy and refactor to have a `SinglerequestReader` + `MultirequestReader`, instead of the current Reader + _SeekableRawReader?

Just thinking here how to keep this module maintainable and readable.

File: smart_open/s3.py [hunk 2]
```diff
@@ -605,31 +695,68 @@ def read(self, size=-1):
...
-                logger.warning(
-                    '%s: caught %r while reading %d bytes, sleeping %ds before retry',
-                    self,
-                    err,
-                    size,
-                    seconds,
-                )
-                time.sleep(seconds)
+        binary_collected = []
+        bytes_read = 0  # == sum(map(len, binary_collected))
```
Comment (ddelange) [inline smart_open/s3.py:699]: could use `binary_collected = bytearray()` to which you can efficiently `+=`, see also https://github.com/piskvorky/smart_open/pull/711

it also has a `len` so you can get rid of bytes_read

File: smart_open/s3.py [hunk 3]
```diff
@@ -738,8 +737,7 @@ def read(self, size=-1):
 
             if binary:
                 self._position += len(binary)
-                binary_collected.append(binary)
-                bytes_read += len(binary)
+                binary_collected += binary
```
Comment (piskvorky) [inline smart_open/s3.py:740]: Not a fan. The previous version was more explicit & simpler IMO.

Is it really guaranteed `bytearray +=` is efficient and doesn't keep re-allocating internally?

Comment (piskvorky) [reply smart_open/s3.py:740]: So, logarithmic reallocations. Not as good as the original `join()` which is constant, but not terrible I guess.

Comment (ddelange) [reply smart_open/s3.py:740]: according to [this article](https://pythonspeed.com/articles/bytesio-reduce-memory-usage/), our best bet seems to be `BytesIO.get_value()`

File: smart_open/s3.py [hunk 4]
```diff
@@ -605,31 +695,66 @@ def read(self, size=-1):
...
+
+            if binary:
                 self._position += len(binary)
-                return binary
+                binary_collected += binary
+            else:
+                # By construction we always try to read >= 1 bytes, so if we see this
+                # we know that we are at the end of the current _body.
+                # here just in case.
+                if self._chunk_end is None or is_fresh_chunk:
```
Comment (ddelange) [inline smart_open/s3.py:745]: both `_chunk_end` and `is_fresh_chunk` variables only exist for this check. judging by the code comments, this _might_ not be necessary at all.

maybe we can construct moto tests to prove that we can avoid this added complexity?

**Comment by ddelange** [comment]
how about this simplification, eliminating `is_fresh_chunk` and `self._chunk_end` (also from `__init__`)?

- Replaces while True with explicit exit conditions
- Removes for-else construct
- Uses BytesIO instead of bytearray

```py
def read(self, size=-1):
    """Read from the continuous connection with the remote peer."""
    inf = float('inf')
    retry_errors = (
        ConnectionResetError,
        botocore.exceptions.BotoCoreError,
        urllib3.exceptions.HTTPError,
    )

    if size < -1:
        raise ValueError(f'size must be >= -1, got {size}')

    if size == -1:
        size = inf

    binary_collected = BytesIO()

    while binary_collected.tell() < size and self._position < self._content_length:
        if self._body is None:
            self._open_body()

        binary = None
        for seconds in [1, 2, 4, 8, 16]:
            try:
                if size == inf:
                    binary = self._body.read()
                else:
                    binary = self._body.read(size - binary_collected.tell())
                break  # success - exit retry loop
            except retry_errors as err:
                logger.warning(
                    '%s: caught %r while reading %d bytes, sleeping %ds before retry',
                    self,
                    err,
                    size,
                    seconds,
                )
                time.sleep(seconds)
                self._open_body()

        if binary is None:
            raise IOError('%s: failed to read after %d attempts' % (self, 5))

        if not binary:
            break  # end of stream - no more data available

        self._position += len(binary)
        binary_collected.write(binary)

    return binary_collected.getvalue()
```

**Comment by batterseapower** [comment]
I don't think that CI failure is real ("The action 'Start vsftpd' has timed out after 2 minutes.")
